### PR TITLE
Arrow Key Cell Switcher

### PR DESCRIPTION
**Pull request by stefnotch**
Implements a quick cell switcher. Works with the two code editors.

However, I have no idea how to make it work with Prosemirror.

### CODE REVIEW

File: src/components/editor/contentEditor.ts [hunk 1]
```diff
@@ -14,35 +15,28 @@ type ProsemirrorModule = Awaited<typeof prosemirrorPromise>;
...
-    @property({type: Object})
-    content: ContentContainer;
+    private cell: Cell;
+    private runtime: Runtime;
 
     createRenderRoot() {
         return this;
     }
 
-    constructor(content: ContentContainer = {textContent: ""}, opts: {focusAfterInit?: boolean} = {}) {
```
Comment (gzuidhof) [inline src/components/editor/contentEditor.ts:38]: The reason I abstracted it to this type is because it makes it more portable (it may also make sense to do the same for the code editors later on, I wish I did it from the start). Smaller interfaces makes for more portable code.

So we could define some sort of interface of stuff we need to make something *editable*, like this:
```typescript
// We can rename this to something better: `TextContentEditable`?
interface ContentContainer {
  textContent: string;
}
```
The current `Cell` object would satisfy this interface, but importantly we can now also re-use it as a text editor for something else! So let's say at some point we want a text editor for the metadata of a cell we could use it for that too, or if I'm making a plugin that embeds multiple markdown editors I could reuse this component for that.

To listen for assignments to the `textContent` field we use a [ES6 proxy](https://github.com/gzuidhof/starboard-notebook/blob/master/src/components/helpers/cellProxy.ts#L44-L58). That way we can act on changes from the text editor if we need to.

To solidify the relationship we could put `implements ContentContainer` on our `Cell` type.

Long story short: let's keep the arg here as `ContentContainer` (or a better name!).

File: src/components/editor/contentEditor.ts [hunk 2]
```diff
@@ -67,7 +61,7 @@ export class StarboardContentEditor extends LitElement {
 
         prosemirrorPromise.then(pm => {
             if (this.view) {
-                this.view.updateState(pm.createEditorState({content: this.content}));
+                // this.view.updateState(pm.createEditorState({content: this.content})); // TODO: dunno
```
Comment (gzuidhof) [inline src/components/editor/contentEditor.ts:64]: 🤔 Yes I think this not necessary, good catch!

File: src/components/editor/prosemirror/module.ts [hunk 3]
```diff
@@ -21,20 +23,39 @@ export { EditorView, EditorState, Plugin, defaultMarkdownSerializer };
...
-                view: () => {
-                    return {
-                        update: debounce((view: EditorView) => {
-                            opts.content.textContent = defaultMarkdownSerializer.serialize(view.state.doc);
-                        }, 50)
-                    };
-                },
-            }),
-        ],
+export function createProseMirrorEditor(element: HTMLElement, cell: Cell, opts: { focusAfterInit?: boolean }, runtime: Runtime) {
```
Comment (gzuidhof) [inline src/components/editor/prosemirror/module.ts:26]: Nit: as the amount of arguments grows, I usually end up putting stuff more and more into an options object as the last argument, could you swap it to be the last? I usually also give it a default value of `= {}` if everything is optional.

As mentioned in the other comment, we can replace `cell` here with the more generic `something with a textContent field` type.

File: src/components/editor/prosemirror/module.ts [hunk 4]
```diff
@@ -21,20 +23,39 @@ export { EditorView, EditorState, Plugin, defaultMarkdownSerializer };
...
+    let editorView = new EditorView(element, {
+        state: EditorState.create({
+            doc: parser.parse(cell.textContent),
+            plugins: [
+                keymap({
+                    "ArrowDown": function (state, dispatch, view) {
+                        if(state.selection.empty) {
+                            // Now what?
```
Comment (gzuidhof) [inline src/components/editor/prosemirror/module.ts:34]: Don't let this be a blocker for merging this PR, I'm happy to merge it without us figuring this out first - we can leave this stub code for now and figure out prosemirror later.

File: src/runtime/create.ts [hunk 5]
```diff
@@ -114,7 +114,7 @@ export function setupRuntime(notebook: StarboardNotebookElement): Runtime {
       notebook.performUpdate();
     },
 
-    runCell(id: string, focusNext: boolean, insertNewCell: boolean) {
+    runCell(id: string, focusNext?: "before" | "after", insertNewCell?: boolean) {
```
Comment (gzuidhof) [inline src/runtime/create.ts:117]: Ah `focusNext` now implies `next` as in `after` time-wise. Kind of confusing (I can't think of a much better name though, `focusAfterRun` or just `focus` maybe?

File: src/components/editor/codeMirror.ts [hunk 6]
```diff
@@ -173,12 +172,45 @@ export function createCodeMirrorEditor(element: HTMLElement, cell: Cell, opts: {
...
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return false; 
+                }
+            }
+        },
+        { key: "ArrowDown", run: (target) => { 
+                if(target.state.selection.ranges.length == 1 && target.state.selection.ranges[0].empty) {
```
Comment (gzuidhof) [inline src/components/editor/codeMirror.ts:192]:
```suggestion
                if(target.state.selection.ranges.length === 1 && target.state.selection.ranges[0].empty) {
```

File: src/components/editor/contentEditor.ts [hunk 7]
```diff
@@ -14,35 +15,28 @@ type ProsemirrorModule = Awaited<typeof prosemirrorPromise>;
...
-    content: ContentContainer;
+    private cell: Cell;
+    private runtime: Runtime;
 
     createRenderRoot() {
         return this;
     }
 
-    constructor(content: ContentContainer = {textContent: ""}, opts: {focusAfterInit?: boolean} = {}) {
+    constructor(cell: Cell, runtime: Runtime, opts: {focusAfterInit?: boolean} = {}) {
```
Comment (gzuidhof) [inline src/components/editor/contentEditor.ts:32]:
```suggestion
    constructor(contentContainer: ContentContainer, runtime: Runtime, opts: {focusAfterInit?: boolean} = {}) {
```

File: src/components/editor/contentEditor.ts [hunk 8]
```diff
@@ -14,35 +15,28 @@ type ProsemirrorModule = Awaited<typeof prosemirrorPromise>;
...
         super();
-        this.content = content;
+        this.runtime = runtime;
+        this.cell = cell;
         
         prosemirrorPromise.then(pm => {
-            this.view = new pm.EditorView(this, {
-                state: pm.createEditorState({content: this.content}),
-            });
+            this.view = pm.createProseMirrorEditor(this, this.cell, opts as any, this.runtime)
```
Comment (gzuidhof) [inline src/components/editor/contentEditor.ts:38]:
```suggestion
            this.view = pm.createProseMirrorEditor(this, this.contentContainer, opts, this.runtime)
```

File: src/components/editor/codeMirror.ts [hunk 9]
```diff
@@ -173,12 +172,45 @@ export function createCodeMirrorEditor(element: HTMLElement, cell: Cell, opts: {
...
+                    let cursorPosition = target.state.selection.ranges[0].head;
+                    if(firstLine.from <= cursorPosition && cursorPosition <= firstLine.to) {
+                        _runtime.controls.emit({id: cell.id, type: "RUN_CELL", focusNext: "before"});
+                        return true
+                    } else {
+                        return false;
+                    }
+                } else {
+                    return false; 
+                }
```
Comment (gzuidhof) [inline src/components/editor/codeMirror.ts:188]: Nit: You could replace with to `return false` without any `elses`. A matter of style but I think it's slightly easier to read.

File: src/components/editor/codeMirror.ts [hunk 10]
```diff
@@ -173,12 +172,45 @@ export function createCodeMirrorEditor(element: HTMLElement, cell: Cell, opts: {
...
+            }
+        },
+        { key: "ArrowDown", run: (target) => { 
+                if(target.state.selection.ranges.length == 1 && target.state.selection.ranges[0].empty) {
+                    let lastline = target.state.doc.line(target.state.doc.lines);
+                    let cursorPosition = target.state.selection.ranges[0].head;
+                    if(lastline.from <= cursorPosition && cursorPosition <= lastline.to) {
+                        _runtime.controls.emit({id: cell.id, type: "RUN_CELL", focusNext: "after"});
```
Comment (gzuidhof) [inline src/components/editor/codeMirror.ts:196]: Wait, do we actually want to `run` the cell or just focus the next? 
We can introduce a new message type `FOCUS_CELL` or something to just shift focus.


Also the `focusEditor` could take an optional argument:

```typescript
focusEditor(opts?: {at?: "start" | "end} = {}) {
  //...
}
 ```
But that can of course be future work outside of this PR :)

File: src/runtime/create.ts (hunk 5 above)
Comment (stefnotch) [reply src/runtime/create.ts]: Okay, I renamed it to `focus` with `previous` and `next` as possible values. I hope that's fine :)

File: src/components/editor/codeMirror.ts (hunk 10 above)
Comment (stefnotch) [reply src/components/editor/codeMirror.ts]: That's a good point, I haven't considered that. I'm still purely thinking about reactive notebooks xD
Fixed it :)

**Comment by stefnotch** [comment]
Thank you so much for the in-depth review! I always super appreciate it when someone helps me improve my code :)
I think I addressed everything with the commit, let me know if I missed anything

File: src/components/editor/monaco.ts [hunk 11]
```diff
@@ -98,10 +98,43 @@ function addEditorKeyboardShortcuts(
         contextMenuGroupId: 'starboard',
         contextMenuOrder: 1,
         run: (_ed) => emit({
-            id: cellId, type: "RUN_CELL", focusNextCell: true, insertNewCell: false
+            id: cellId, type: "RUN_CELL", focus: "next"
         })
     });
 
+    editor.addAction({
+        id: 'run-cell-and-next',
```
Comment (gzuidhof) [inline src/components/editor/monaco.ts:106]: rename (also below) to something like "Focus next cell"

**Review by gzuidhof — APPROVED** [review]
Thank you so much :).
I added one last small comment, ready to merge after that!

File: src/components/editor/monaco.ts (hunk 11 above)
Comment (stefnotch) [reply src/components/editor/monaco.ts]: Okay, fixed!
